File: tests/ccl_test_growth.c
#include "ccl.h"
#include "ctest.h"
#include <stdio.h>
#include <math.h>

// The tolerance in D(z) for all the
#define GROWTH_TOLERANCE 1.0e-4

CTEST_DATA(growth) {
  double Omega_c;
  double Omega_b;
  double h;
  double A_s;
  double n_s;
  double N_nu_rel;
  double N_nu_mass;
  double mnu;
  double Omega_v[5];
  double Omega_k[5];
  double w_0[5];
  double w_a[5];
  
  double z[6];
  double gf[5][6];
};

// Read the fixed format file containing all the growth factor
// benchmarks
static void read_growth_test_file(double z[6], double gf[5][6])
{
  //Growth is normalized to ~a at early times
  FILE * f = fopen("./tests/benchmark/growth_model1-5.txt", "r");
  ASSERT_NOT_NULL(f);
  
  // Ignore header line
  char str[1024];
  char* rtn;
  rtn = fgets(str, 1024, f);
  
    // File is fixed format - five rows and six columns
  for (int i=0; i<6; i++) {
    int count = fscanf(f, "%le %le %le %le %le %le\n", &z[i],
		       &gf[0][i], &gf[1][i], &gf[2][i], &gf[3][i], &gf[4][i]);
    	// Check that all the stuff in the benchmark is there
    ASSERT_EQUAL(6, count);
  }
  fclose(f);
}

// Set up the cosmological parameters to be used in each of the
// models
CTEST_SETUP(growth) {
  // Values that are the same for all 5 models
  data->Omega_c = 0.25;
  data->Omega_b = 0.05;
  data->h = 0.7;
  data->A_s = 2.1e-9;
  data->n_s = 0.96;
  data->N_nu_rel=0;
  data->N_nu_mass=0;
  data->mnu=0;
  
  
  // Values that are different for the different models
  double Omega_v[5] = {  0.7,  0.7,  0.7,  0.65, 0.75 };
  double w_0[5]     = { -1.0, -0.9, -0.9, -0.9, -0.9  };
  double w_a[5]     = {  0.0,  0.0,  0.1,  0.1,  0.1  };
  
  // Fill in the values from these constant arrays.
  for (int i=0; i<5; i++) {
    data->Omega_v[i] = Omega_v[i];
    data->w_0[i]     = w_0[i];
    data->w_a[i]     = w_a[i];
    data->Omega_k[i] = 1.0 - data->Omega_c - data->Omega_b - data->Omega_v[i];
  }

  // The file of benchmark data.
  read_growth_test_file(data->z, data->gf);
}

static void compare_growth(int model, struct growth_data * data)
{
  int status=0; 	
  // Make the parameter set from the input data
  // Values of some parameters depend on the model index
  ccl_parameters params = ccl_parameters_create(data->Omega_c, data->Omega_b, data->Omega_k[model], data->N_nu_rel, data->N_nu_mass, data->mnu, data->w_0[model], data->w_a[model], data->h, data->A_s, data->n_s,-1,NULL,NULL, &status);
  params.Omega_g=0;
  // Make a cosmology object from the parameters with the default configuration
  ccl_cosmology * cosmo = ccl_cosmology_create(params, default_config);
  ASSERT_NOT_NULL(cosmo);
  
  // Compare to benchmark data
  for (int j=0; j<6; j++) {
    double a = 1/(1.+data->z[j]);
    double gf_ij=ccl_growth_factor_unnorm(cosmo,a, &status);
    if (status) printf("%s\n",cosmo->status_message);
    double absolute_tolerance = GROWTH_TOLERANCE*data->gf[model][j];
    if (fabs(absolute_tolerance)<1e-12) absolute_tolerance = 1e-12;
    ASSERT_DBL_NEAR_TOL(data->gf[model][j], gf_ij, absolute_tolerance);
  }

  ccl_cosmology_free(cosmo);
}

//This test code compares the modified growth function computed by CCL
//against the exact result for a particular modification of the growth rate.
static void check_mgrowth(void)
{
  int ii,nz_mg=128;
  double *z_mg,*df_mg;
  ccl_parameters params1,params2;
  ccl_cosmology *cosmo1,*cosmo2;
  int status=0;
  z_mg=malloc(nz_mg*sizeof(double));
  df_mg=malloc(nz_mg*sizeof(double));
  for(ii=0;ii<nz_mg;ii++) {
    z_mg[ii]=4*(ii+0.0)/(nz_mg-1.);
    df_mg[ii]=0.1/(1+z_mg[ii]);
  }
  params1=ccl_parameters_create(0.25,0.05,0,0,0,0,-1,0,0.7,2.1E-9,0.96,-1,NULL,NULL, &status);
  params2=ccl_parameters_create(0.25,0.05,0,0,0,0,-1,0,0.7,2.1E-9,0.96,nz_mg,z_mg,df_mg, &status);
  cosmo1=ccl_cosmology_create(params1,default_config);
  cosmo2=ccl_cosmology_create(params2,default_config);

  //We have included a growth modification \delta f = K*a (with K==0.1 arbitrarily)
  //This case has an analytic solution, given by D(a) = D_0(a)*exp(K*(a-1))
  //Here we check the growth computed by the library with the analytic solution.
  for(ii=0;ii<nz_mg;ii++) {
    double a=1./(1+z_mg[ii]);
    double d1=ccl_growth_factor(cosmo1,a,&status);
    double d2=ccl_growth_factor(cosmo2,a,&status);
    double f1=ccl_growth_rate(cosmo1,a,&status);
    double f2=ccl_growth_rate(cosmo2,a,&status);
    double f2r=f1+0.1*a;
    double d2r=d1*exp(0.1*(a-1));
    ASSERT_DBL_NEAR_TOL(d2r/d2,1.,GROWTH_TOLERANCE);
    ASSERT_DBL_NEAR_TOL(f2r/f2,1.,GROWTH_TOLERANCE);
  }

  free(z_mg);
  free(df_mg);
  ccl_cosmology_free(cosmo1);
  ccl_cosmology_free(cosmo2);
}

CTEST2(growth, model_1) {
  int model = 0;
  compare_growth(model, data);
}

CTEST2(growth, model_2) {
  int model = 1;
  compare_growth(model, data);
}

CTEST2(growth, model_3) {
  int model = 2;
  compare_growth(model, data);
}

CTEST2(growth, model_4) {
  int model = 3;
  compare_growth(model, data);
}

CTEST2(growth, model_5) {
  int model = 4;
  compare_growth(model, data);
}

CTEST2(growth,mgrowth) {
  check_mgrowth();
}

File: tests/ccl_test_growth.py
import numpy as np
from numpy.testing import assert_allclose, run_module_suite
import pyccl as ccl
from os.path import dirname,join
# Set tolerances
GROWTH_TOLERANCE = 1e-4

# Set up the cosmological parameters to be used in each of the models
# Values that are the same for all 5 models
Omega_c = 0.25
Omega_b = 0.05
N_nu_rel = 0.
N_nu_mass = 0.
m_nu=0.
h = 0.7
A_s = 2.1e-9
n_s = 0.96

# Values that are different for the different models
Omega_v_vals = np.array([0.7, 0.7, 0.7, 0.65, 0.75])
w0_vals = np.array([-1.0, -0.9, -0.9, -0.9, -0.9])
wa_vals = np.array([0.0, 0.0, 0.1, 0.1, 0.1])

def read_growth_test_file():
    """
    Read the file containing all the radial comoving distance benchmarks 
    (distances are in Mpc/h)
    """
    # Load data from file
    dat = np.genfromtxt(join(dirname(__file__),"benchmark/growth_model1-5.txt")).T
    assert(dat.shape == (6,6))
    
    # Split into redshift column and growth(z) columns
    z = dat[0]
    gfac = dat[1:]
    return z, gfac

# Set-up test data
z, gfac = read_growth_test_file()

def compare_growth(z, gfac_bench, Omega_v, w0, wa):
    """
    Compare growth factor calculated by pyccl with the values in the benchmark 
    file. This test only works if radiation is explicitly set to 0.
    """
    # Set Omega_K in a consistent way
    Omega_k = 1.0 - Omega_c - Omega_b - Omega_v    
    
    # Create new Parameters and Cosmology objects
    p = ccl.Parameters(Omega_c=Omega_c, Omega_b=Omega_b, N_nu_rel=N_nu_rel, N_nu_mass=N_nu_mass, m_nu=m_nu,
                       h=h, A_s=A_s, n_s=n_s, Omega_k=Omega_k,
                       w0=w0, wa=wa)
    p.parameters.Omega_g = 0. # Hack to set to same value used for benchmarks
    cosmo = ccl.Cosmology(p)
    
    # Calculate distance using pyccl
    a = 1. / (1. + z)
    gfac = ccl.growth_factor_unnorm(cosmo, a)
    
    # Compare to benchmark data
    assert_allclose(gfac, gfac_bench, atol=1e-12, rtol=GROWTH_TOLERANCE)


def test_growth_model_0():
    i = 0
    compare_growth(z, gfac[i], Omega_v_vals[i], w0_vals[i], wa_vals[i])

def test_growth_model_1():
    i = 1
    compare_growth(z, gfac[i], Omega_v_vals[i], w0_vals[i], wa_vals[i])

def test_growth_model_2():
    i = 2
    compare_growth(z, gfac[i], Omega_v_vals[i], w0_vals[i], wa_vals[i])

def test_growth_model_3():
    i = 3
    compare_growth(z, gfac[i], Omega_v_vals[i], w0_vals[i], wa_vals[i])

def test_growth_model_4():
    i = 4
    compare_growth(z, gfac[i], Omega_v_vals[i], w0_vals[i], wa_vals[i])


def test_mgrowth():
    """
    Compare the modified growth function computed by CCL against the exact 
    result for a particular modification of the growth rate.
    """
    # Define differential growth rate arrays
    nz_mg = 128
    z_mg = np.zeros(nz_mg)
    df_mg = np.zeros(nz_mg)
    for i in range(0, nz_mg):
        z_mg[i] = 4. * (i + 0.0) / (nz_mg - 1.)
        df_mg[i] = 0.1 / (1. + z_mg[i])
    
    # Define two test cosmologies, without and with modified growth respectively
    p1 = ccl.Parameters(Omega_c=0.25, Omega_b=0.05, Omega_k=0., N_nu_rel=0., N_nu_mass=0., m_nu=0., 
                        w0=-1., wa=0., h=0.7, A_s=2.1e-9, n_s=0.96)
    p2 = ccl.Parameters(Omega_c=0.25, Omega_b=0.05, Omega_k=0., N_nu_rel=0., N_nu_mass=0., m_nu=0.,  
                        w0=-1., wa=0., h=0.7, A_s=2.1e-9, n_s=0.96, 
                        z_mg=z_mg, df_mg=df_mg)
    cosmo1 = ccl.Cosmology(p1)
    cosmo2 = ccl.Cosmology(p2)
    
    # We have included a growth modification \delta f = K*a, with K==0.1 
    # (arbitrarily). This case has an analytic solution, given by 
    # D(a) = D_0(a)*exp(K*(a-1)). Here we compare the growth computed by CCL
    # with the analytic solution.
    a = 1. / (1. + z_mg)
    
    d1 = ccl.growth_factor(cosmo1, a)
    d2 = ccl.growth_factor(cosmo2, a)
    f1 = ccl.growth_rate(cosmo1, a)
    f2 = ccl.growth_rate(cosmo2, a)
    
    f2r = f1 + 0.1*a
    d2r = d1 * np.exp(0.1*(a-1.))
    
    # Check that ratio of calculated and analytic results is within tolerance
    assert_allclose(d2r/d2, np.ones(d2.size), rtol=GROWTH_TOLERANCE)
    assert_allclose(f2r/f2, np.ones(f2.size), rtol=GROWTH_TOLERANCE)


if __name__ == "__main__":
    run_module_suite()
